src/1-theory/10-async.adoc

```diff
@@ -0,0 +1,407 @@
+t-shirts. We created the whole thing over beers, and we made some whopping
+mistakes. Our payment acceptance sequence looked like the synchronous one above, as
+we didn't care about speed at that early stage. The biggest mistake was that
+when the payment API gave us a positive response to the `POST` request, we
+immediately considered the payment good and marked the shirts for dispatch.
+
+It turns out there was a web hook we were meant to listen out for, and this
+would confirm if the payment had been accepted, or not... Seeing as we totally
+ignored that, there were a few t-shirts that went out for free! Oops.
+
+This was simple enough, we added a `/callback` endpoint, registered it with the
+payment gateway, and they would fire updates at it a few seconds or minutes
+after a payment attempt was made. We simply marked orders as pending in the mean
+time, made a order status page, and they would get an email later. We could have
+added 30 second short polling on the status page if we cared, or just had the
+page refresh, but it was midnight and the bartender had just opened the whiskey.
+
+Web hooks are often thought of as a bit of a dark art, because they happen outside of the normal
+request/response stuff that API developers are used to. It can be hard to see them happen,
+tricky to confirm they are sending what you think, and used to be tricky to document.
+
+These days there are loads of awesome tools around to help. OpenAPI v3.0 added support for
+https://swagger.io/docs/specification/callbacks/[Callbacks].
+
+NOTE: It makes me sad that this is called callbacks when it should be called web
+hooks, and it's on the swagger.io website when the specification has been renamed
+to OpenAPI. Ahh well.
+
+Tools to help with this stuff:
+
+- https://pipedream.com/[PipeDream] - Formerly known as RequestBin, create an endpoint and register that as the callback URL, and see what traffic is coming in. Then maybe copy that to use in a test suite
+- https://ngrok.com/[ngrok] - Create a tunnel to your locally running application and register your local app as the callback URL
+
+That ngrok example might have made you notice an issue with web hooks: you need
+some sort of DNS or known IP address to fire the payload at. If the client is
+written in PHP, Java, Node, Ruby, etc. and running behind a web server then
+yeah, it will probably have one of those, but if the client is a JavaScript
+application running in the users browser then that isn't going to work.
+
+For that we need another approach, and that is... 🥁
+
+=== WebSocket
+
+WebSocket is a whole other protocol to HTTP, that operates over TCP directly.
+Instead of a client asking the server over and over again, or making a blocking
+connection to the web server, the client establishes a connection to the server,
+and then the two parties can send messages back and forth in real-time, and the
+server doesn't need to know the IP address of the client because the client
+initiated the connection. Cool!
```

Comment:
well actually, it does know the IP address because they have a connection :-P
it is clear what you mean but still not quite correct :)

Reply:
It knows the public IP of the interface making the connection, so maybe the router or some shit, but it wont know the IP of the machine. Should I get into that?

Reply:
Not really, but the point of this is still not clear (for me it is clear because I know how these work) I just think that it would be hard to understand for someone not familiar with it.

Above you say that we will not open a connection for too long to save valuable time on the webserver, instead we introduce callbacks/webhooks to solve that. Now we get to WebSockets, which does exactly what we wanted to avoid, opening a connection for quite some (probably very long) time :)

```diff
Instead of a client asking the server over and over again, or making a blocking
-connection to the web server, the client establishes a connection to the server,
+connection to the web server, the client establishes a connection to another server which runs a websocket daemon,
and then the two parties can send messages back and forth in real-time, and the
-server doesn't need to know the IP address of the client because the client
+server doesn't need to know the IP address for creating new connections when there is something to send, because the client
initiated the connection. Cool!
```

Reply:
Awesome points, you're totally right. I've addeda TODO for this to hit in the second draft.